core/src/runtime/syscall.rs

```diff
@@ -69,13 +70,19 @@ impl SyscallCode {
   }
 }
 
+pub enum Outcome {
+  Result(Option<u32>),
+  Exit(u32),
+}
+
+pub(crate) type SyscallResult = Result<Outcome, StatusCode>;
+
 pub trait Syscall: Send + Sync {
-  /// Execute the syscall and return the resulting value of register a0. `arg1` and `arg2` are the
-  /// values in registers X10 and X11, respectively. While not a hard requirement, the convention
```

Comment:
No, the "convention" is unchanged - a syscall might return a result (hence the `Result(Option<u32>)` variant of the `Outcome` enum). I only changed the `execute` trait function to:
- allow returning errors (instead of panicking) - hence it returns a `Result<Outcome, StatusCode>`
- avoid doing dynamic dispatch in the Runtime and having special meaning fields in the `SyscallContext` (like the `exit_code`) that were only used in a specific syscall (HALT). Instead, the "you should exit" outcome is just one of the variants in the `Outcome` enum.